From: Kuan-Wei Chiu <visitorckw@gmail.com>
Subject: lib min_heap: add typedef for sift_down function pointer
Date: Wed, 11 Jun 2025 05:55:10 +0800

Several min-heap operations such as min_heapify_all(), min_heap_pop(),
min_heap_pop_push(), and min_heap_del() use min_heap_sift_down()
internally.  With the addition of the equal-elements-aware variant
min_heap_sift_down_eqaware(), these functions now need to choose between
multiple sift_down implementations.

Introduce a siftdown_fn_t typedef to represent the function pointer type
for sift_down routines.  This avoids repeating verbose function pointer
declarations and simplifies code that dynamically selects the appropriate
implementation based on heap characteristics.

Link: https://lkml.kernel.org/r/20250610215516.1513296-3-visitorckw@gmail.com
Signed-off-by: Kuan-Wei Chiu <visitorckw@gmail.com>
Cc: Ching-Chun (Jim) Huang <jserv@ccns.ncku.edu.tw>
Cc: Coly Li <colyli@kernel.org>
Cc: Jonathan Corbet <corbet@lwn.net>
Cc: Kent Overstreet <kent.overstreet@linux.dev>
Cc: Robert Pang <robertpang@google.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 include/linux/min_heap.h |    3 +++
 1 file changed, 3 insertions(+)

--- a/include/linux/min_heap.h~lib-min_heap-add-typedef-for-sift_down-function-pointer
+++ a/include/linux/min_heap.h
@@ -49,6 +49,9 @@ struct min_heap_callbacks {
 	void (*swp)(void *lhs, void *rhs, void *args);
 };
 
+typedef void (*siftdown_fn_t)(min_heap_char *heap, size_t pos, size_t elem_size,
+			    const struct min_heap_callbacks *func, void *args);
+
 /**
  * is_aligned - is this pointer & size okay for word-wide copying?
  * @base: pointer to data
_
